### PR TITLE
fix data-driven split handedness pin assignement

### PR DESCRIPTION
**Pull request by Kriechi**
## Description

_(updated PR description based on changes from discussion)_

This PR fixes a number of issues with the current data-driven configuration for split handedness detection:
* There was a mix and inconsistent usage of `split.primary` and `split.main`. The generator for `config.h` was looking for a `primary` key, the but jsonschema and `info.py` only used `main`.
* `matrix_grid` config was incorrectly generated with additional `{ }` around `GP1,GP2` pin pairs
* handedness by pin high or low never was hooked up - missing pin value
* naming weirdness of `main`/`primary` - which has nothing to do with detecting which side we are on
* missing options for left or right configs when using pin or matrix_grid depending on low/high state when reading the pin or grid intersection

This PR marks the old json object `split.main` as deprecated while remaining a backwards compatible jsonschema, and introduces a new object for side detection:
```
"split": {
    "handedness": {
        "method": "pin|matrix_grid|eeprom|usb",
        "pin": "GP1",                          // only needed when using pin method
        "matrix_grid": ["GP1", "GP2"],         // only needed when using matrix_grid method
        "determined_side": "left|right"        // optional, only used with pin, matrix_grid, or fixed method
    }
}
```

This PR brings the config generation, the info command, and the only existing usage of the now deprecated data-driven config in line with the new `side_detection` schema and handles the conversion of old to new options.

This data-driven config maps to the following config.h keys:
```
#define SPLIT_HAND_PIN GP1
#define SPLIT_HAND_PIN_LOW_IS_LEFT

#define SPLIT_HAND_MATRIX_GRID GP1, GP2
#define SPLIT_HAND_MATRIX_GRID_LOW_IS_RIGHT

#define EE_HANDS

#define MASTER_LEFT
#define MASTER_RIGHT
```

## Types of Changes

- [X] Core
- [X] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by sigprof** [comment]
The more important question is why `main` is even used here? The handedness options do not select which side of the keyboard is “main”/“master”, they select which side is left (but the default method is driven by the master detection status). Only the `"main": "left"` and `"main": "right"` values make some sense.

The handedness configuration should probably look like this:
```json
"split": {
    "side_detection": {
        "method": "pin",
        "pin": "GP4",
        "low_side": "left"
    }
}
```
```json
"split": {
    "side_detection": {
        "method": "matrix_grid",
        "matrix_grid": ["GP4", "GP10"],
        "low_side": "left"
    }
}
```
```json
"split": {
    "side_detection": {
        "method": "eeprom"
    }
}
```
```json
"split": {
    "side_detection": {
        "method": "main_status",
        "main_side": "left"
    }
}
```
To avoid confusing defaults, we could make the `low_side` parameter mandatory when `"method": "pin"` or `"method": "matrix_grid"` is specified in JSON (currently at the C level `SPLIT_HAND_PIN` defaults to `"low_side": "right"`, but `SPLIT_HAND_MATRIX_GRID` defaults to `"low_side": "left"`); alternatively we could do a breaking change for `SPLIT_HAND_PIN`, but that might be a lot of work.

**Comment by Kriechi** [comment]
Thanks @sigprof - this makes way more sense!

I've pushed the changes - with small changes to your proposed naming.
I did not add strict validation / error throwing for weird combinations, with the idea that somebody might want to configure "pin" method, while keeping the line for matrix_grid for debugging or other reasons. So whatever the method is set to - will only read the fields that are actually needed by the currently selected method.

And I think I got away with all the LOW/HIGH default handling without too much issues - please verify.

File: lib/python/qmk/cli/generate/config_h.py [hunk 1]
```diff
@@ -127,22 +127,28 @@ def generate_encoder_config(encoder_json, config_h_lines, postfix=''):
...
-                config_h_lines.append('#        define MASTER_LEFT')
-            elif kb_info_json['split']['primary'] == 'right':
-                config_h_lines.append('#        define MASTER_RIGHT')
-            config_h_lines.append('#    endif // MASTER_RIGHT')
+            config_h_lines.append('#   ifndef MASTER_RIGHT')
+            if kb_info_json['split']['side_detection']['determined_side'] == 'left':
+                config_h_lines.append('#       define MASTER_LEFT')
+            elif kb_info_json['split']['side_detection']['determined_side'] == 'right':
+                config_h_lines.append('#       define MASTER_RIGHT')
+            config_h_lines.append('#   endif // MASTER_RIGHT')
```
Comment (fauxpark) [inline lib/python/qmk/cli/generate/config_h.py:139]:
```suggestion
            config_h_lines.append('#    ifndef MASTER_RIGHT')
            if kb_info_json['split']['side_detection']['determined_side'] == 'left':
                config_h_lines.append('#        define MASTER_LEFT')
            elif kb_info_json['split']['side_detection']['determined_side'] == 'right':
                config_h_lines.append('#        define MASTER_RIGHT')
            config_h_lines.append('#    endif // MASTER_RIGHT')
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: lib/python/qmk/cli/generate/config_h.py (hunk 1 above)
Comment (Kriechi) [reply lib/python/qmk/cli/generate/config_h.py]: Thanks - included now!

**Comment by daskygit** [comment]
There's been some further discussion on [discord](https://discord.com/channels/440868230475677696/867530715192885269/1022614573011963924) with no resolution.

Quick recap is I'd like to see the options renamed with `method: usb' requiring it's own further configuration option. For example,

```json
"split": {
    "handedness": {
        "method": "pin|matrix_grid|eeprom|usb",
        "pin": "GP1",
        "matrix_grid": ["GP1", "GP2"],
        "active_low": "left|right",
        "usb_side": "left|right"
    }
}
```

File: data/schemas/keyboard.jsonschema [hunk 2]
```diff
@@ -443,7 +443,15 @@
...
             "properties": {
                 "enabled": {"type": "boolean"},
+                "main": {
+                    "$comment": "deprecated - see side_detection instead",
+                    "deprecated": true,
+                    "type": "string",
+                    "enum": ["eeprom", "left", "matrix_grid", "pin", "right"]
+                },
                 "matrix_grid": {
+                    "$comment": "deprecated - see side_detection instead",
```
Comment (evils) [inline data/schemas/keyboard.jsonschema:453]: seems like that maybe got renamed?

```suggestion
                    "$comment": "deprecated - see handedness instead",
                    "deprecated": true,
                    "type": "string",
                    "enum": ["eeprom", "left", "matrix_grid", "pin", "right"]
                },
                "matrix_grid": {
                    "$comment": "deprecated - see handedness instead",
```

Comment (Kriechi) [reply data/schemas/keyboard.jsonschema]: Yes! Thanks!

**Comment by Kriechi** [comment]
bump - please assign *awaiting review* label, or close directly.

**Comment by Kriechi** [comment]
Partially implemented and superseeded by #22369.

**Comment by zvecr** [comment]
Closing as I plan to re-implement based on the current state of the repo.

**Pull request closed without merging**